### ISSUE TITLE
generated bedrock2 code says load_word when load 1 byte is indended

### ISSUE THREAD

**Issue by andres-erbsen** [mit-plv/fiat-crypto#771]
in `from_bytes` this causes incorrect values to be loaded (and probably undefined behavior at loading the last word). (from https://github.com/mit-plv/bedrock2/pull/155#issue-411559079)

**Comment by jadephilipoom** [comment]
I've looked into this and I think I've figured out the problem and what to do about it, so I figured I'd infodump here.

Basically, the issue is that the bedrock2 translation, in the same vein as expecting every variable to be full-width (e.g. 64-bit), also expects input/output arrays to always be arrays of full-width words. Therefore, from_bytes would need to take in an array of 64-bit words, each of which represented one byte. This is obviously not the type signature we'd want for that function. Instead, in-memory arrays should be exempt from the full-width expectation.

The fix for this will involve making the bedrock2 translation aware of the sizes of its inputs/outputs (it has no intermediate in-memory arrays), by passing in to the top level of the translation a `Syntax.access_size` for each array. That information can be threaded through to make the correct load/store instruction. So the code will be fine; on to the proofs.

One complexity is now that the equivalence relation for in-memory arrays needs to be changed to take these bounds into account. This relation's inputs are a fiat-crypto AST representation (`list Z` in this case) and a bedrock2 implementation (a pointer to an in-memory array). It will now need to know that none of the `Z`s are too big for the access size. We can get that information from the input/output `list_Z_bounded_by` proofs, which will now be preconditions for the bedrock2 translation.

Long story short, it's fixable, but the fact that it requires modifying the equivalence relation on which the whole translation depends will be a bit of a pain. I'm considering it my highest-priority task and will hopefully have a draft fix soon.

**Comment by andres-erbsen** [comment]
I think the appropriate parameter for the representation relation might be `access_size`. and if the relation always used word.unsigned instead of word. of_Z (assuming unsigned code), it needn't even know that it is only going to handle inputs that are bounded as desired.

**Comment by andres-erbsen** [comment]
yeah, seems so.

 I'd like to check the status of a long-time feature request of mine: what is currently the nicest way to iterate on the definition of an arithmetic function and get to see it's compilation with some arguments? My bar for actually jumping in and trying to fix that is being able to do this iteration within one coq file, I think. it's entirely plausible that what I'm asking for is already there and I'm just not aware where.

**Comment by jadephilipoom** [comment]
> if the relation always used word.unsigned instead of word. of_Z (assuming unsigned code), it needn't even know that it is only going to handle inputs that are bounded as desired.

In the case of this equivalence relation, the left hand side is `API.interp t` for some type in the fiat-crypto AST, which evaluates to `list Z` for the type corresponding to lists. So I don't see a way I can avoid the `word.of_Z` and take in `list Semantics.word` instead -- I'd have to write some kind of function that crawls over the fiat-crypto output and converts to words, but that would just move the `word.of_Z` to a different place.

> I'd like to check the status of a long-time feature request of mine: what is currently the nicest way to iterate on the definition of an arithmetic function and get to see it's compilation with some arguments? My bar for actually jumping in and trying to fix that is being able to do this iteration within one coq file, I think. it's entirely plausible that what I'm asking for is already there and I'm just not aware where.

You mean, you want to plug in a Gallina function and see the output after reification + rewriter? The [old version of the bedrock2 backend tests](https://github.com/mit-plv/fiat-crypto/blob/4adcee9b60b98642e13c080005d4df9076697120/src/Bedrock/Tests/X25519_64.v) are possibly a good example of how to do that (they start from the Gallina definitions derived in `Arithmetic/ModOps.v` -- the new ones go through `PushButtonSynthesis/UnsaturatedSolinas.v`, so they're not as straightforward).

**Comment by andres-erbsen** [comment]
I do think you can avoid word.of_Z, at least given what I understand of the situation. you can still decode the words and then say they are equal to the interpretation of the expression. up to you, though.

the second post was meant for the other bug...

**Comment by jadephilipoom** [comment]
> I do think you can avoid word.of_Z, at least given what I understand of the situation. you can still decode the words and then say they are equal to the interpretation of the expression. up to you, though.

The problem is, where do I get these words from? I need to state that the fiat-crypto AST interpretation is equivalent to the bedrock2 interpretation. The fiat-crypto AST interpretation (`API.interp_type _`) is in terms of `Z`, not words, so I can't really change that signature to be words instead. Are you suggesting I change the definition of `API.interp_type`? Or I suppose I could put an existential into the equivalence, something like `fun (lhs : list Z) (rhs : _) => exists ws, map word.unsigned ws = lhs ...` But that's pretty useless; it'll be `exists (map word.of_Z lhs)` every time.

Anyway, I still don't see how avoiding `word.of_Z` in a useful way is possible. If you have a concrete suggestion I'm all ears.

**Comment by andres-erbsen** [comment]
it's the existential that I'm suggesting. if you don't want to do it, don't do it.
